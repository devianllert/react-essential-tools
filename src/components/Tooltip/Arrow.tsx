import styled from 'styled-components';

interface ArrowProps {
  placement: string;
}

interface ArrowPlacement {
  [key: string]: string;
}

const arrowPlacement: ArrowPlacement = {
  bottom: `
    top: 0;
    left: 0;
    margin-top: -4px;
    height: 8px;
    width: 4px;

    &::before {
      border-width: 0 4px 4px 4px;
      border-color: transparent transparent currentcolor transparent;
    }
  `,
  top: `
    bottom: 0;
    left: 0;
    margin-bottom: -4px;
    height: 8px;
    width: 4px;

    &::before {
      border-width: 4px 4px 0 4px;
      border-color: currentcolor transparent transparent transparent;
    }
  `,
  right: `
    left: 0;
    margin-left: -4px;
    height: 8px;
    width: 4px;

    &::before {
      border-width: 4px 4px 4px 0;
      border-color: transparent currentcolor transparent transparent;
    }
  `,
  left: `
    right: 0;
    margin-right: -4px;
    height: 8px;
    width: 4px;

    &::before {
      border-width: 4px 0 4px 4px;
      border-color: transparent transparent transparent currentcolor;
    }
  `,
};


export const Arrow = styled.span<ArrowProps>`
  position: absolute;
  font-size: 6px;
  color: #000;

  &::before {
    content: "";
    margin: auto;
    display: block;
    width: 0;
    height: 0;
    border-style: solid;
  }

  ${(props): string => arrowPlacement[props.placement]}
`;
