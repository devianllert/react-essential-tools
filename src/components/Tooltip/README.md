# `Tooltip`

Tooltips display informative text when users hover over, focus on, or tap an element.

## Example

```jsx
import React from 'react';
import { Tooltip } from 'react-essential-tools';

export default function SimpleTooltip() {
  return (
    <Tooltip title="tooltip">
      <span>Text with tooltip</span>
    </Tooltip>
  );
}
```

## Custom child element

The tooltip needs to apply DOM event listeners to its child element. If the child is a custom React element, you need to make sure that it spreads its properties to the underlying DOM element.

```jsx
function MyComponent(props) {
  //  Spread the properties to the underlying DOM element.
  return <div {...props}>Bin</div>
}

// ...

<Tooltip title="tooltip">
  <MyComponent />
</Tooltip>
```