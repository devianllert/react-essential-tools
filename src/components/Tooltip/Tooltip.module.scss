/* Styles applied to the Popper component. */
.popper {
  z-index: 15;
  pointer-events: none;
}
/* Styles applied to the Popper component if `interactive={true}`. */
.popperInteractive {
  pointer-events: auto;
}
/* Styles applied to the tooltip (label wrapper) element. */
.tooltip {
  background-color: #000;
  border-radius: 4px;
  color: #fff;
  font-family: "Roboto", sans-serif;
  padding: 4px 8px;
  font-size: 12px;
  max-width: 300px;
  word-wrap: break-word;
}
/* Styles applied to the tooltip (label wrapper) element if the tooltip is opened by touch. */
.touch {
  padding: 8px 16px;
  font-size: 14px;
}
/* Styles applied to the tooltip (label wrapper) element if `placement` contains "left". */
.tooltipPlacementLeft {
  transform-origin: right center;
  margin: 0 4px;
}
/* Styles applied to the tooltip (label wrapper) element if `placement` contains "right". */
.tooltipPlacementRight {
  transform-origin: left center;
  margin: 0 4px;
}
/* Styles applied to the tooltip (label wrapper) element if `placement` contains "top". */
.tooltipPlacementTop {
  transform-origin: center bottom;
  margin: 4px 0;
}
/* Styles applied to the tooltip (label wrapper) element if `placement` contains "bottom". */
.tooltipPlacementBottom {
  transform-origin: center top;
  margin: 4px 0;
}