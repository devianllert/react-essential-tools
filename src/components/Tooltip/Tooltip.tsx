import React, {
  useState,
  useCallback,
  useEffect,
  useRef,
  cloneElement,
  ReactElement,
  ChangeEvent,
} from 'react';
import clsx from 'clsx';

import { useIsFocusVisible } from '../../utils/focusVisible';

import { Popper } from '../Popper';

import styles from './Tooltip.module.scss';

interface Props {
  children: React.ReactElement;
  disableFocusListener?: boolean;
  disableHoverListener?: boolean;
  disableTouchListener?: boolean;
  enterDelay?: number;
  enterTouchDelay?: number;
  id?: string;
  interactive?: boolean;
  leaveDelay?: number;
  leaveTouchDelay?: number;
  onClose?: (event: ChangeEvent<{}>) => void;
  onOpen?: (event: ChangeEvent<{}>) => void;
  open?: boolean;
  placement?:
    | 'bottom-end'
    | 'bottom-start'
    | 'bottom'
    | 'left-end'
    | 'left-start'
    | 'left'
    | 'right-end'
    | 'right-start'
    | 'right'
    | 'top-end'
    | 'top-start'
    | 'top';
  PopperProps?: Partial<import('../Popper/Popper').Props>;
  title: React.ReactNode;
}

const capitalize = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);

export const Tooltip = (props: Props): ReactElement => {
  const {
    children,
    disableFocusListener = false,
    disableHoverListener = false,
    disableTouchListener = false,
    enterDelay = 0,
    enterTouchDelay = 700,
    id,
    interactive = false,
    leaveDelay = 0,
    leaveTouchDelay = 1500,
    onClose,
    onOpen,
    open: openProp,
    placement = 'bottom',
    PopperProps,
    title,
    ...other
  } = props;

  const { isFocusVisible, onBlurVisible, ref: focusVisibleRef } = useIsFocusVisible();

  const [openState, setOpenState] = useState(false);
  const [childIsFocusVisible, setChildIsFocusVisible] = useState(false);
  const [, forceUpdate] = useState(0);
  const [childNode, setChildNode] = useState<Element>();

  const ignoreNonTouchEvents = useRef(false);
  const { current: isControlled } = useRef(openProp != null);

  const defaultId = useRef<string>();
  const closeTimer = useRef<number>();
  const enterTimer = useRef<number>();
  const leaveTimer = useRef<number>();
  const touchTimer = useRef<number>();

  const handleRef = useCallback(
    (instance: HTMLElement) => {
      setChildNode(instance);
      focusVisibleRef(instance);
    },
    [focusVisibleRef],
  );

  useEffect(() => {
    // Fallback to this default id when possible.
    // Use the random value for client-side rendering only.
    // We can't use it server-side.
    if (!defaultId.current) {
      defaultId.current = `tooltip-${Math.round(Math.random() * 1e5)}`;
    }

    // Rerender with defaultId and childNode.
    if (openProp) {
      forceUpdate((n) => n + 1);
    }
  }, [openProp]);

  useEffect(
    () => (): void => {
      clearTimeout(closeTimer.current);
      clearTimeout(enterTimer.current);
      clearTimeout(leaveTimer.current);
      clearTimeout(touchTimer.current);
    },
    [],
  );

  const handleOpen = (event: ChangeEvent): void => {
    // The mouseover event will trigger for every nested element in the tooltip.
    // We can skip rerendering when the tooltip is already open.
    // We are using the mouseover event instead of the mouseenter event to fix a hide/show issue.
    if (!isControlled && !openState) {
      setOpenState(true);
    }

    if (onOpen) {
      onOpen(event);
    }
  };

  const handleEnter = (event: ChangeEvent): void => {
    const childrenProps = children.props;

    if (event.type === 'mouseover' && childrenProps.onMouseOver) {
      childrenProps.onMouseOver(event);
    }

    if (ignoreNonTouchEvents.current && event.type !== 'touchstart') {
      return;
    }

    // Remove the title ahead of time.
    // We don't want to wait for the next render commit.
    // We would risk displaying two tooltips at the same time (native + this one).
    if (childNode) {
      childNode.removeAttribute('title');
    }

    clearTimeout(enterTimer.current);
    clearTimeout(leaveTimer.current);

    if (enterDelay) {
      event.persist();

      enterTimer.current = window.setTimeout(() => {
        handleOpen(event);
      }, enterDelay);
    } else {
      handleOpen(event);
    }
  };

  const handleBlur = (): void => {
    if (childIsFocusVisible) {
      setChildIsFocusVisible(false);
      onBlurVisible();
    }
  };

  const handleFocus = (event: ChangeEvent): void => {
    // Workaround for https://github.com/facebook/react/issues/7769
    // The autoFocus of React might trigger the event before the componentDidMount.
    // We need to account for this eventuality.
    if (!childNode) {
      setChildNode(event.currentTarget);
    }

    if (isFocusVisible(event)) {
      setChildIsFocusVisible(true);
      handleEnter(event);
    }

    const childrenProps = children.props;
    if (childrenProps.onFocus) {
      childrenProps.onFocus(event);
    }
  };

  const handleClose = (event: ChangeEvent): void => {
    if (!isControlled) {
      setOpenState(false);
    }

    if (onClose) {
      onClose(event);
    }

    clearTimeout(closeTimer.current);

    closeTimer.current = window.setTimeout(() => {
      ignoreNonTouchEvents.current = false;
    }, 150);
  };

  const handleLeave = (event: ChangeEvent): void => {
    const childrenProps = children.props;

    if (event.type === 'blur') {
      if (childrenProps.onBlur) {
        childrenProps.onBlur(event);
      }
      handleBlur();
    }

    if (event.type === 'mouseleave' && childrenProps.onMouseLeave) {
      childrenProps.onMouseLeave(event);
    }

    clearTimeout(enterTimer.current);
    clearTimeout(leaveTimer.current);

    event.persist();

    leaveTimer.current = window.setTimeout(() => {
      handleClose(event);
    }, leaveDelay);
  };

  const handleTouchStart = (event: ChangeEvent): void => {
    ignoreNonTouchEvents.current = true;
    const childrenProps = children.props;

    if (childrenProps.onTouchStart) {
      childrenProps.onTouchStart(event);
    }

    clearTimeout(leaveTimer.current);
    clearTimeout(closeTimer.current);
    clearTimeout(touchTimer.current);

    event.persist();

    touchTimer.current = window.setTimeout(() => {
      handleEnter(event);
    }, enterTouchDelay);
  };

  const handleTouchEnd = (event: ChangeEvent): void => {
    if (children.props.onTouchEnd) {
      children.props.onTouchEnd(event);
    }

    clearTimeout(touchTimer.current);
    clearTimeout(leaveTimer.current);

    event.persist();

    leaveTimer.current = window.setTimeout(() => {
      handleClose(event);
    }, leaveTouchDelay);
  };

  let open: boolean = isControlled ? openProp || false : openState;

  // There is no point in displaying an empty tooltip.
  if (title === '') {
    open = false;
  }

  const shouldShowNativeTitle = !open && !disableHoverListener;

  const childrenProps = {
    'aria-describedby': open ? id || defaultId.current : null,
    title: shouldShowNativeTitle && typeof title === 'string' ? title : null,
    ...other,
    ...children.props,
  };

  if (!disableTouchListener) {
    childrenProps.onTouchStart = handleTouchStart;
    childrenProps.onTouchEnd = handleTouchEnd;
  }

  if (!disableHoverListener) {
    childrenProps.onMouseOver = handleEnter;
    childrenProps.onMouseLeave = handleLeave;
  }

  if (!disableFocusListener) {
    childrenProps.onFocus = handleFocus;
    childrenProps.onBlur = handleLeave;
  }

  const interactiveWrapperListeners = interactive
    ? {
      onMouseOver: childrenProps.onMouseOver,
      onMouseLeave: childrenProps.onMouseLeave,
      onFocus: childrenProps.onFocus,
      onBlur: childrenProps.onBlur,
    }
    : {};

  return (
    <>
      {cloneElement(children, { ref: handleRef, ...childrenProps })}

      <Popper
        className={clsx(styles.popper, {
          [styles.popperInteractive]: interactive,
        })}
        placement={placement}
        anchorEl={childNode}
        open={childNode ? open : false}
        id={childrenProps['aria-describedby']}
        transition
        {...interactiveWrapperListeners} // eslint-disable-line
        {...PopperProps} // eslint-disable-line
      >
        {({ placement: placementInner }): ReactElement => (
          <div
            className={clsx(
              styles.tooltip,
              {
                [styles.touch]: ignoreNonTouchEvents.current,
              },
              styles[`tooltipPlacement${capitalize(placementInner.split('-')[0])}`],
            )}
          >
            {title}
          </div>
        )}
      </Popper>
    </>
  );
};
