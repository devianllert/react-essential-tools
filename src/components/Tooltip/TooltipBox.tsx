import styled from 'styled-components';

interface TooltipBoxProps {
  interactive?: boolean;
  arrow?: boolean;
  placement: string;
}

interface TransformPlacement {
  [key: string]: string;
}

const transformPlacement: TransformPlacement = {
  bottom: `
    transform-origin: center top;
    margin: 4px 0;
  `,
  top: `
    transform-origin: center bottom;
    margin: 4px 0;
  `,
  right: `
    transform-origin: left center;
    margin: 0 4px;
  `,
  left: `
    transform-origin: right center;
    margin: 0 4px;
  `,
};

export const TooltipBox = styled.div<TooltipBoxProps>`
  position: relative
  background-color: #000;
  border-radius: 4px;
  color: #fff;
  font-family: "Roboto", sans-serif;
  padding: ${(props): string => (props.interactive ? '8px 16px' : '4px 8px')};
  font-size: ${(props): string => (props.interactive ? '14px' : '12px')};
  max-width: 300px;
  word-wrap: break-word;

  ${(props): string => transformPlacement[props.placement]}
`;
