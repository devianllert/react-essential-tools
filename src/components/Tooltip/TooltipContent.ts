import styled from 'styled-components';

interface TooltipContentProps {
  interactive: boolean;
  arrow: boolean;
  placement: string;
}

interface TransformPlacement {
  [key: string]: string;
}

export const TooltipContent = styled.div<TooltipContentProps>`
  position: relative;
  background-color: #000;
  border-radius: 4px;
  color: #fff;
  font-family: "Roboto", sans-serif;
  padding: ${({ interactive }): string => (interactive ? '8px 16px' : '4px 8px')};
  font-size: ${({ interactive }): string => (interactive ? '14px' : '12px')};
  max-width: 300px;
  word-wrap: break-word;

  transform-origin: ${({ placement }): string => ({
    bottom: 'center top',
    top: 'center bottom',
    right: 'left center',
    left: 'right center',
  } as TransformPlacement)[placement]};

  margin: ${({ placement, arrow }): string => ({
    bottom: arrow ? '6px 0' : '4px 0',
    top: arrow ? '6px 0' : '4px 0',
    right: arrow ? '0 6px' : '0 4px',
    left: arrow ? '0 6px' : '0 4px',
  } as TransformPlacement)[placement]};
`;
