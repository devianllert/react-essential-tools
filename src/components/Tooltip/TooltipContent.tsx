import styled from 'styled-components';

interface TooltipContentProps {
  interactive: boolean;
  arrow: boolean;
  placement: string;
}

interface TransformPlacement {
  [key: string]: string;
}

export const TooltipContent = styled.div<TooltipContentProps>`
  position: relative;
  background-color: #000;
  border-radius: 4px;
  color: #fff;
  font-family: "Roboto", sans-serif;
  padding: ${(props): string => (props.interactive ? '8px 16px' : '4px 8px')};
  font-size: ${(props): string => (props.interactive ? '14px' : '12px')};
  max-width: 300px;
  word-wrap: break-word;

  transform-origin: ${(props): string => ({
    bottom: 'center top',
    top: 'center bottom',
    right: 'left center',
    left: 'right center',
  } as TransformPlacement)[props.placement]};

  margin: ${(props): string => ({
    bottom: props.arrow ? '6px 0' : '4px 0',
    top: props.arrow ? '6px 0' : '4px 0',
    right: props.arrow ? '0 6px' : '0 4px',
    left: props.arrow ? '0 6px' : '0 4px',
  } as TransformPlacement)[props.placement]};
`;
