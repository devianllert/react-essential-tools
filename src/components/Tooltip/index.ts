export { Tooltip } from './Tooltip';
export * from './TooltipArrow';
