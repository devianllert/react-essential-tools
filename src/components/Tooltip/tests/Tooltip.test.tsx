import React from 'react';
import {
  render,
  fireEvent,
  cleanup,
  act,
} from '@testing-library/react';

import { Tooltip } from '../Tooltip';

import '@testing-library/jest-dom/extend-expect';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: (): void => {},
        scheduleUpdate: (): void => {},
      };
    }
  };
});

jest.useFakeTimers();

describe('<Tooltip />', () => {
  afterEach(cleanup);

  const defaultProps = {
    children: <span data-testid="children">Hello World</span>,
    title: 'test tooltip',
  };

  it('should render the correct structure', () => {
    const { getByTestId } = render(
      <Tooltip title={defaultProps.title}>
        {defaultProps.children}
      </Tooltip>,
    );

    expect(getByTestId('children')).toBeDefined();

    expect(getByTestId('children').hasAttribute('title')).toBeTruthy();
    expect(getByTestId('children').getAttribute('title')).toEqual('test tooltip');

    expect(getByTestId('children').textContent).toEqual('Hello World');
  });

  it('should render tooltip when hover', () => {
    const { getByTestId } = render(
      <Tooltip title={defaultProps.title}>
        {defaultProps.children}
      </Tooltip>,
    );

    expect(document.querySelector('[role="tooltip"]')).not.toBeInTheDocument();


    act(() => {
      fireEvent.mouseOver(getByTestId('children'));
    });

    expect(document.querySelector('[role="tooltip"]')).toBeInTheDocument();
    expect(document.querySelector('[role="tooltip"]')!.textContent).toEqual('test tooltip');
  });

  it('should unmount tooltip when mouse leave', () => {
    const { getByTestId } = render(
      <Tooltip title={defaultProps.title}>
        {defaultProps.children}
      </Tooltip>,
    );

    expect(document.querySelector('[role="tooltip"]')).not.toBeInTheDocument();

    fireEvent.mouseOver(getByTestId('children'));

    expect(document.querySelector('[role="tooltip"]')).toBeInTheDocument();

    fireEvent.mouseLeave(getByTestId('children'));

    act(() => {
      jest.runAllTimers();
    });

    expect(document.querySelector('[role="tooltip"]')).not.toBeInTheDocument();
  });

  describe('prop: title', () => {
    it('should display if the title is present', () => {
      render(
        <Tooltip title={defaultProps.title} open>
          {defaultProps.children}
        </Tooltip>,
      );

      expect(document.querySelector('[role="tooltip"]')).toBeInTheDocument();
    });

    it('should not display if the title is an empty string', () => {
      render(
        <Tooltip title="" open>
          {defaultProps.children}
        </Tooltip>,
      );

      expect(document.querySelector('[role="tooltip"]')).not.toBeInTheDocument();
    });

    it('should be passed down to the child as a native title', () => {
      const { getByRole } = render(
        <Tooltip title="Hello World">
          <button type="submit">Hello World</button>
        </Tooltip>,
      );

      expect(getByRole('button').getAttribute('title')).toEqual('Hello World');
    });
  });

  describe('snapshots', () => {
    it('should render and match the snapshot', () => {
      const { container } = render(
        <Tooltip title="Hello World" id="tooltip">
          <button type="submit">Hello World</button>
        </Tooltip>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render and match the snapshot when hovered', () => {
      const { container, getByRole } = render(
        <Tooltip title="Hello World" id="tooltip">
          <button type="submit">Hello World</button>
        </Tooltip>,
      );

      fireEvent.mouseEnter(getByRole('button'));

      expect(container).toMatchSnapshot();
    });

    it('should render and match the tooltip snapshot', () => {
      const { getByRole } = render(
        <Tooltip title="Hello World" id="tooltip">
          <button type="submit">Hello World</button>
        </Tooltip>,
      );

      fireEvent.mouseEnter(getByRole('button'));

      expect(document.querySelector('[role="tooltip"]')).toMatchSnapshot();
    });
  });
});
