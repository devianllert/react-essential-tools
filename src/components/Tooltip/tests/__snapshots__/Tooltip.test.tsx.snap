// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> snapshots should render and match the snapshot 1`] = `
<div>
  <button
    title="Hello World"
    type="submit"
  >
    Hello World
  </button>
</div>
`;

exports[`<Tooltip /> snapshots should render and match the snapshot when hovered 1`] = `
<div>
  <button
    aria-describedby="tooltip"
    type="submit"
  >
    Hello World
  </button>
</div>
`;

exports[`<Tooltip /> snapshots should render and match the tooltip snapshot 1`] = `
<div
  class="popper"
  id="tooltip"
  role="tooltip"
  style="position: fixed;"
>
  <div
    class="tooltip tooltipPlacementBottom"
  >
    Hello World
  </div>
</div>
`;
